pkg/tso/keyspace_group_manager.go

```diff
@@ -305,218 +334,29 @@ func (kgm *KeyspaceGroupManager) Close() {
-			return
-		}
-
-		keyspaceGroupsLoaded += uint32(len(groups))
-
-		if watchStartRevision == 0 || revision < watchStartRevision {
-			watchStartRevision = revision
-		}
-
-		// Update the keyspace groups
-		for _, group := range groups {
-			select {
-			case <-ctx.Done():
-				err = errs.ErrLoadKeyspaceGroupsTerminated
-				return
-			default:
-			}
-
-			if group.ID == mcsutils.DefaultKeyspaceGroupID {
-				defaultKGConfigured = true
-			}
-
-			kgm.updateKeyspaceGroup(group)
-		}
-
-		if !more {
-			break
-		}
-	}
-
-	log.Info("loaded keyspace groups", zap.Uint32("keyspace-groups-loaded", keyspaceGroupsLoaded))
-	return
-}
-
-// loadKeyspaceGroups loads keyspace groups from the start ID with limit.
-// If limit is 0, it will load all keyspace groups from the start ID.
-func (kgm *KeyspaceGroupManager) loadKeyspaceGroups(
-	ctx context.Context, startID uint32, limit int64,
-) (revision int64, ksgs []*endpoint.KeyspaceGroup, more bool, err error) {
-	rootPath := kgm.legacySvcRootPath
-	startKey := strings.Join([]string{rootPath, endpoint.KeyspaceGroupIDPath(startID)}, "/")
-	endKey := strings.Join(
-		[]string{rootPath, clientv3.GetPrefixRangeEnd(endpoint.KeyspaceGroupIDPrefix())}, "/")
-	opOption := []clientv3.OpOption{clientv3.WithRange(endKey), clientv3.WithLimit(limit)}
-
-	var (
-		i    int
-		resp *clientv3.GetResponse
-	)
-	for ; i < kgm.loadFromEtcdMaxRetryTimes; i++ {
-		resp, err = etcdutil.EtcdKVGet(kgm.etcdClient, startKey, opOption...)
-
-		failpoint.Inject("delayLoadKeyspaceGroups", func(val failpoint.Value) {
-			if sleepIntervalSeconds, ok := val.(int); ok && sleepIntervalSeconds > 0 {
-				time.Sleep(time.Duration(sleepIntervalSeconds) * time.Second)
-			}
-		})
-
-		failpoint.Inject("loadKeyspaceGroupsTemporaryFail", func(val failpoint.Value) {
-			if maxFailTimes, ok := val.(int); ok && i < maxFailTimes {
-				err = errors.New("fail to read from etcd")
-				failpoint.Continue()
-			}
-		})
-
-		if err == nil && resp != nil {
-			break
-		}
-
-		select {
-		case <-ctx.Done():
-			return 0, []*endpoint.KeyspaceGroup{}, false, errs.ErrLoadKeyspaceGroupsTerminated
-		case <-time.After(defaultLoadFromEtcdRetryInterval):
-		}
-	}
-
-	if i == kgm.loadFromEtcdMaxRetryTimes {
-		return 0, []*endpoint.KeyspaceGroup{}, false, errs.ErrLoadKeyspaceGroupsRetryExhausted.FastGenByArgs(err)
-	}
-
-	kgs := make([]*endpoint.KeyspaceGroup, 0, len(resp.Kvs))
-	for _, item := range resp.Kvs {
-		kg := &endpoint.KeyspaceGroup{}
-		if err = json.Unmarshal(item.Value, kg); err != nil {
-			return 0, nil, false, err
-		}
-		kgs = append(kgs, kg)
-	}
-
-	if resp.Header != nil {
-		revision = resp.Header.Revision + 1
-	}
-
-	return revision, kgs, resp.More, nil
-}
-
-// startKeyspaceGroupsMetaWatchLoop repeatedly watches any change in keyspace group membership/distribution
-// and apply the change dynamically.
-func (kgm *KeyspaceGroupManager) startKeyspaceGroupsMetaWatchLoop(revision int64) {
+func (kgm *KeyspaceGroupManager) startRetryUpdateLoop() {
```

Comment:
No need to add startRetryUpdateLoop(). Can we define postEventFn, similar to deleteFn and putFn, and pass it to watchLoop to execute after "for _, event := range wresp.Events {" loop and before revision = wresp.Header.Revision + 1 in LooperWatch.Watch? This can keep the generality of the watcher and also keep the exact same logic.

Reply:
good idea

Reply:
```
	postEventFn := func() {
	       // Retry the groups that are not initialized successfully before.
	       for id, group := range kgm.groupUpdateRetryList {
		        delete(kgm.groupUpdateRetryList, id)
		        kgm.updateKeyspaceGroup(group)
	        }
	}
```